Sources/libhostmgr/libhostmgr.swift

```diff
@@ -0,0 +1,312 @@
+        try await unpackVM(name: name)
+        return try await startVM(name: name)
+    }
+
+    let destination = FileManager.default.temporaryFilePath(named: name + ".tmp.pvm")
+    try FileManager.default.removeItemIfExists(at: destination)
+    try FileManager.default.copyItem(at: sourceVM.path, to: destination)
+    Console.info("Created temporary VM at \(destination)")
+
+    guard let parallelsVM = try Parallels().importVM(at: destination)?.asStoppedVM() else {
+        Console.crash(message: "Unable to import VM: \(destination)", reason: .unableToImportVM)
+    }
+
+    Console.success("Successfully Imported \(parallelsVM.name) with UUID \(parallelsVM.uuid)")
+
+    try applyVMSettings([
+        .memorySize(Int(ProcessInfo().physicalMemory / 1024 / 1024) - 4096),  // We should make this configurable
+        .cpuCount(ProcessInfo().physicalProcessorCount),
+        .hypervisorType(.apple),
+        .networkType(.shared),
+        .isolateVM(.on),
+        .sharedCamera(.off)
+    ], to: parallelsVM)
+
+    try parallelsVM.start(wait: false)
+
+    let _: Void = try await withCheckedThrowingContinuation { continuation in
+        do {
+            try waitForVMStartup(parallelsVM)
+        } catch {
+            continuation.resume(with: .failure(error))
+        }
+
+        let elapsed = Date().timeIntervalSince(startDate)
+        Console.success("Booted \(parallelsVM.name) \(Format.time(elapsed))")
+        continuation.resume(with: .success(()))
+    }
+}
+
+public func stopAllRunningVMs(
+    immediately: Bool = true,
+    parallelsRepository: ParallelsVMRepositoryProtocol = ParallelsVMRepository()
+) throws {
+    for parallelsVM in try parallelsRepository.lookupRunningVMs() {
+        try stopRunningVM(name: parallelsVM.name, immediately: immediately)
+    }
+}
+
+public func stopRunningVM(
+    name: String,
+    immediately: Bool,
+    parallelsRepository: ParallelsVMRepositoryProtocol = ParallelsVMRepository()
+) throws {
+    let parallelsVM = try lookupParallelsVMOrExit(withIdentifier: name)
+
+    guard let vmToStop = parallelsVM.asRunningVM() else {
+        Console.exit(
+            message: "\(parallelsVM.name) is not running, so it can't be stopped",
+            style: .warning
+        )
+    }
+
+    Console.info("Shutting down \(parallelsVM.name)")
+
+    try vmToStop.shutdown(immediately: immediately)
+    try vmToStop.unregister()
+
+    // Clean up after ourselves by deleting the VM files
+    let vmPath = FileManager.default.temporaryDirectory.appendingPathComponent(parallelsVM.name + ".tmp.pvm")
+
+    guard try FileManager.default.directoryExists(at: vmPath) else {
+        Console.success("Shutdown Complete")
+        return
+    }
+
+    Console.info("Deleting VM storage from \(vmPath)")
+    try FileManager.default.removeItem(at: vmPath)
+
+    Console.success("Shutdown Complete")
+}
+
+func waitForVMStartup(_ parallelsVirtualMachine: StoppedVM) throws {
+    repeat {
+        usleep(100)
```

Comment:
Not a change request, but my gut instinct is checking every 1/1000 is too frequent for VM booting (which takes maybe around 30 seconds?).

Reply:
It's a good instinct! Doing this isn't much better than just not sleeping at all. Adjusted in 21d4cd4dc6a21fb4396d4dc45db8b93ea76416cf